src/main/scala/com/databricks/spark/sql/perf/mllib/BenchmarkAlgorithm.scala

```diff
@@ -0,0 +1,94 @@
+package com.databricks.spark.sql.perf.mllib
+
+import com.typesafe.scalalogging.slf4j.Logging
+
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.evaluation.Evaluator
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
+
+/**
+ * The description of a benchmark for an ML algorithm. It follows a simple, standard proceduce:
+ *  - generate some test and training data
+ *  - generate a model against the training data
+ *  - score the model against the training data
+ *  - score the model against the test data
+ *
+ * You should not assume that your implementation can carry state around. If some state is needed,
+ * consider adding it to the context.
+ *
+ * It is assumed that the implementation is going to be an object.
+ */
+trait BenchmarkAlgorithm extends Logging {
+
+  // This is internal to the implementation of the benchmark.
+  type Model <: Transformer
```

Comment:
I don't think it will need to.  For a specific alg, that alg's test should set the output columns and override the Evaluator setup as needed, so the generic transform() and evaluate() calls should handle everything.